controllers/appwrapper_controller.go

```diff
@@ -110,56 +101,80 @@ func (r *AppWrapperReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 		klog.Error(err, "unable to fetch appwrapper")
 	}
 
-	factory := machineinformersv1beta1.NewSharedInformerFactoryWithOptions(machineClient, resyncPeriod()(), machineinformersv1beta1.WithNamespace(""))
-	informer := factory.Machine().V1beta1().MachineSets().Informer()
-	msLister = factory.Machine().V1beta1().MachineSets().Lister()
-	machineLister = factory.Machine().V1beta1().Machines().Lister()
+	// Adds finalizer to the appwrapper if it doesn't exist
+	if !controllerutil.ContainsFinalizer(&appwrapper, finalizerName) && appwrapper.Status.State != "Completed" {
```

Comment:
Just a suggestion to use constants such as the existing `arbv1.AppWrapperStateCompleted`, instead of hardwritten states.